File: Languages/python/algorithms-dsa/leetcode/trees/lowest_common_ancestor.py
"""
LeetCode #236 - Lowest Common Ancestor of a Binary Tree
Topic: Trees
Difficulty: Medium

Find LCA using recursive post-order traversal.

Time Complexity: O(n)
Space Complexity: O(h)
"""


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def lowest_common_ancestor(root: TreeNode | None, p: TreeNode, q: TreeNode) -> TreeNode | None:
    if not root or root is p or root is q:
        return root
    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)
    if left and right:
        return root
    return left or right


if __name__ == "__main__":
    # Tree: [3,5,1,6,2,0,8,null,null,7,4]
    n7 = TreeNode(7)
    n4 = TreeNode(4)
    n6 = TreeNode(6)
    n2 = TreeNode(2, n7, n4)
    n5 = TreeNode(5, n6, n2)
    n0 = TreeNode(0)
    n8 = TreeNode(8)
    n1 = TreeNode(1, n0, n8)
    root = TreeNode(3, n5, n1)

    assert lowest_common_ancestor(root, n5, n1) is root
    assert lowest_common_ancestor(root, n5, n4) is n5
    print("All tests passed!")

File: Languages/go/algorithms-dsa/leetcode/trees/lowest_common_ancestor.go
package main

import "fmt"

/*
  LC 236 - Lowest Common Ancestor of a Binary Tree
  Topic: Trees (DFS)
  Difficulty: Medium
  Time: O(n) | Space: O(h)
*/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)
	if left != nil && right != nil {
		return root
	}
	if left != nil {
		return left
	}
	return right
}

func main() {
	//       3
	//      / \
	//     5   1
	//    / \
	//   6   2
	n6 := &TreeNode{6, nil, nil}
	n2 := &TreeNode{2, nil, nil}
	n5 := &TreeNode{5, n6, n2}
	n1 := &TreeNode{1, nil, nil}
	root := &TreeNode{3, n5, n1}

	lca := lowestCommonAncestor(root, n5, n1)
	if lca == root {
		fmt.Println("PASS: LCA(5,1) = 3")
	} else {
		fmt.Println("FAIL: LCA(5,1)")
	}

	lca2 := lowestCommonAncestor(root, n5, n2)
	if lca2 == n5 {
		fmt.Println("PASS: LCA(5,2) = 5")
	} else {
		fmt.Println("FAIL: LCA(5,2)")
	}
}
